vertx-web-graphql/src/main/java/io/vertx/ext/web/handler/graphql/impl/GraphiQLHandlerImpl.java

```diff
@@ -0,0 +1,152 @@
+/*
+ * Copyright 2019 Red Hat, Inc.
+ *
+ * Red Hat licenses this file to you under the Apache License, version 2.0
+ * (the "License"); you may not use this file except in compliance with the
+ * License.  You may obtain a copy of the License at:
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.vertx.ext.web.handler.graphql.impl;
+
+import io.vertx.core.MultiMap;
+import io.vertx.core.http.HttpHeaders;
+import io.vertx.core.http.HttpServerResponse;
+import io.vertx.core.http.impl.MimeMapping;
+import io.vertx.core.impl.logging.Logger;
+import io.vertx.core.impl.logging.LoggerFactory;
+import io.vertx.core.json.JsonObject;
+import io.vertx.ext.web.MIMEHeader;
+import io.vertx.ext.web.RoutingContext;
+import io.vertx.ext.web.handler.graphql.GraphiQLHandler;
+import io.vertx.ext.web.handler.graphql.GraphiQLHandlerOptions;
+import io.vertx.ext.web.impl.Utils;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URL;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Scanner;
+import java.util.function.Function;
+
+/**
+ * @author Thomas Segismont
+ */
+public class GraphiQLHandlerImpl implements GraphiQLHandler {
+
+  private static final Logger log = LoggerFactory.getLogger(GraphiQLHandlerImpl.class);
+
+  private static final Function<RoutingContext, MultiMap> DEFAULT_GRAPHIQL_REQUEST_HEADERS_FACTORY = rc -> null;
+
+  private final GraphiQLHandlerOptions options;
+
+  private Function<RoutingContext, MultiMap> graphiQLRequestHeadersFactory = DEFAULT_GRAPHIQL_REQUEST_HEADERS_FACTORY;
+
+  public GraphiQLHandlerImpl(GraphiQLHandlerOptions options) {
+    Objects.requireNonNull(options, "options");
+    this.options = options;
+  }
+
+  @Override
+  public GraphiQLHandler graphiQLRequestHeaders(Function<RoutingContext, MultiMap> factory) {
+    graphiQLRequestHeadersFactory = factory != null ? factory : DEFAULT_GRAPHIQL_REQUEST_HEADERS_FACTORY;
+    return this;
+  }
+
+  @Override
+  public void handle(RoutingContext rc) {
+    if (!options.isEnabled()) {
+      rc.next();
+      return;
+    }
+    HttpServerResponse response = rc.response();
+    String filename = Utils.pathOffset(rc.normalisedPath(), rc);
+    if (filename.isEmpty()) {
+      if (rc.parsedHeaders().accept().stream().map(MIMEHeader::subComponent).anyMatch(sub -> "html".equalsIgnoreCase(sub))) {
+        rc.response().setStatusCode(301).putHeader(HttpHeaders.LOCATION, rc.currentRoute().getPath()).end();
+      } else {
+        rc.next();
+      }
+      return;
+    }
+    if (filename.equals("/")) {
+      filename = "/index.html";
+    }
+    String resource = loadResource(filename);
```

Comment:
why not load the `index.html` at start and cache, and make this class extend `StaticHandlerImpl` to serve the remaining files using the same way we do with static files? This way all cache directives are handled, etc...

Reply:
I had not thought about reusing the static handler. I will experiment with this idea.

Reply:
Done. I used composition instead of inheritance because exposing all static handler props makes little sense here.